flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeChecks.java

```diff
@@ -332,4 +351,162 @@ public Boolean visit(DayTimeIntervalType dayTimeIntervalType) {
+			}
+			if (thisType.getClass() != thatType.getClass()) {
+				return false;
+			}
+			return thisType.isNullable() == thatType.isNullable()
+				&& thisType.getTypeRoot() == thatType.getTypeRoot();
+		}
+
+		@Override
+		protected Boolean defaultMethod(LogicalType thatType) {
+			if (canEqual(thatType)) {
+				// for atomic types, we use equals simply.
+				return thisType.equals(thatType);
+			} else {
+				return false;
+			}
+		}
+
+		@Override
+		public Boolean visit(RowType thatType) {
+			if (!canEqual(thatType)) {
+				return false;
+			}
+			List<RowType.RowField> thisFields = ((RowType) thisType).getFields();
+			List<RowType.RowField> thatFields = thatType.getFields();
+			if (thisFields.size() != thatFields.size()) {
+				return false;
+			}
+			for (int i = 0; i < thisFields.size(); i++) {
+				RowType.RowField thisField = thisFields.get(i);
+				RowType.RowField thatField = thatFields.get(i);
+				// ignore field names
+				if (!thisField.getType().equals(thatField.getType())
+					|| !Objects.equals(thisField.getDescription(), thatField.getDescription())) {
+					return false;
+				}
+			}
+			return true;
+		}
+
+		@Override
+		public Boolean visit(ArrayType thatType) {
+			if (!canEqual(thatType)) {
+				return false;
+			}
+			LogicalType thisElementType = ((ArrayType) thisType).getElementType();
+			LogicalType thatElementType = thatType.getElementType();
+			return areTypeEqualsWithoutNames(thisElementType, thatElementType);
+		}
+
+		@Override
+		public Boolean visit(MultisetType thatType) {
+			if (!canEqual(thatType)) {
+				return false;
+			}
+			LogicalType thisElementType = ((MultisetType) thisType).getElementType();
+			LogicalType thatElementType = thatType.getElementType();
+			return areTypeEqualsWithoutNames(thisElementType, thatElementType);
+		}
+
+		@Override
+		public Boolean visit(MapType thatType) {
+			if (!canEqual(thatType)) {
+				return false;
+			}
+			LogicalType thisKeyType = ((MapType) thisType).getKeyType();
+			LogicalType thisValueType = ((MapType) thisType).getValueType();
+
+			LogicalType thatKeyType = thatType.getKeyType();
+			LogicalType thatValueType = thatType.getValueType();
+
+			return areTypeEqualsWithoutNames(thisKeyType, thatKeyType) &&
+				areTypeEqualsWithoutNames(thisValueType, thatValueType);
+		}
+
+		@Override
+		public Boolean visit(DistinctType thatType) {
+			if (!canEqual(thatType)) {
+				return false;
+			}
+			boolean thisIsFinal = ((DistinctType) thisType).isFinal();
+			ObjectIdentifier thisObjectIdentifier = ((DistinctType) thisType).getObjectIdentifier();
+			Optional<String> thisDescription = ((DistinctType) thisType).getDescription();
+			LogicalType thisSourceType = ((DistinctType) thisType).getSourceType();
+
+			return thisIsFinal == thatType.isFinal() &&
+				thisObjectIdentifier.equals(thatType.getObjectIdentifier()) &&
+				thisDescription.equals(thatType.getDescription()) &&
+				areTypeEqualsWithoutNames(thisSourceType, thatType.getSourceType());
```

Comment:
I think we can just compare the source type. Everything else is logical meta data, no?